http-webhook.md

```diff
@@ -0,0 +1,339 @@
+a registration. In extreme cases, a notification infrastructure could be abused
+to launch denial-of-service attacks against an arbitrary web-site.
+
+To protect the sender from being abused in such a way, a legitimate delivery
+target needs to indicate that it agrees with notifications being delivered to it.
+
+Reaching the delivery agreement is realized using the following validation
+handshake. The handshake can either be executed immediately at registration
+time or as a "pre-flight" request immediately preceding a delivery. If the
+handshake outcome is that the delivery is denied, the sender MUST NOT deliver
+events to the target.
+
+It is important to understand is that the handshake does not aim to establish
+an authentication or authorization context. It only serves to protect the
+sender from being told to a push to a destination that is not expecting the
+traffic. While this specification mandates use of an authorization model, this
+mandate is not sufficient to protect any arbitrary website from unwanted
+traffic if that website doesn't implement access control and therefore ignores
+the `Authorization` header.
+
+Delivery targets SHOULD support the abuse protection feature. If a target does
+not support the feature, the sender MAY choose not to send to the target, at
+all, or send only at a very low request rate.
+
+### 4.1.1. Validation request
+
+The validation request uses the HTTP [OPTIONS][OPTIONS] method. The request
+is directed to the exact resource target URI that is being registered.
+
+With the validation request, the sender asks the target for permission to
+send notifications, and it can declare a desired request rate (requests per
+minute).
+
+The delivery target will respond with a permission statement and the permitted
+request rate.
+
+The following header fields MUST be included in the validation request:
+
+#### 4.1.1.2. WebHook-Request-Origin
+
+The `WebHook-Request-Origin` header requests permission to send notifications
+from this sender, and contains a DNS expression that identifies the sending
+system, for example "eventemitter.example.com". The value is meant to summarily
+identify all sender instances that act on the behalf of a certain system, and
+not an individual host.
+
+After the handshake and when permission is granted, the sender MUST use the
+`Origin` request header for each delivery request, with the value matching that
+of this header.
+
+For example:
+
+``` text
+WebHook-Request-Origin: eventemitter.example.com
+```
+
+#### 4.1.1.2. WebHook-Request-Rate
+
+The `WebHook-Request-Rate` header MAY be included in the request and asks for
+permission to send notifications from this sender at the specified rate. The
+value is the string representation of a positive integer number greater than
+zero and expresses the request rate in "requests per minute".
+
+For example, the following header asks for permission to send 120 requests per
+minute:
+
+``` text
+WebHook-Request-Rate: 120
+```
```

Comment:
This webhook seems like a reasonable place to negotiate all kinds of features:

1. Valid timeout for processing?
2. Valid data types (e.g. schemas, versioning)
3. Accepted encodings

In general I've held back on these features since we've put the control plane APIs out of scope for now. This PR suddenly puts these (very important) features back in scope.

Reply:
Addressing both comments at once:

First:
1. I cannot use the CORS headers, because this is a very different problem to solve, and I can't interfere with the actual CORS use-case
2. The point of this mechanism is to provide a common path and not document everything everyone does now. Microsoft will have to make changes for this as well.

Second:

Further negotiations could indeed take place in this handshake, and they should lean on HTTP constructs. We could, for instance, have a reversing HTTP Accept handshake here, meaning that the target gets to declare what format it would like. I don't think that's critical for a first stab, but it would be good to play through and think about what to add. We have the CloudEvents versioning strategy and we do have preliminary media types.